File: LinkedListCycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        unordered_map<ListNode *, int> m;
        if(head ==NULL || head->next ==NULL )
            return false;
        while(head != NULL)
        {
            if (m[head]!=1)
                m[head]=1;
            else
                return true;; 
            head=head->next;
        }
        
    return false;
        
    }
};

File: leet-code-solutions/linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        fast=head
        slow=head

        while fast:
            if fast.next:
                fast=fast.next
            else:
                break

                
            if fast.next:
                fast=fast.next
            else:
                break

            slow=slow.next
            if slow == fast:
                return True
                
        return False